### DISCUSSION TITLE
Copy command for script execution

### DISCUSSION THREAD

**Discussion by BambuschJ** [dnGrep/dnGrep#971] [Ideas]
for some operations it is very useful to copy the file edited by dnGrep script into a backup file before.
Is it possible to integrate a command "copy %1 to *.sik" for this purpose?

**Comment by doug24** [comment]
I'm trying to understand what you are looking for.
Is `%1` a single file?  Maybe I haven't thought about it enough, but what script action will produce one and only one file? Or is it a set of files?
The `to *.sik` argument: does this mean copy the file(s) to the current directory and change the extension to `sik`?

As you probably know, the existing copyfiles command will copy the files in the search results to another directory, without changing the extension.

**Comment by BambuschJ** [comment]
Hello Doug P,

I have posted my actual solution in the attachment.

current I/we work as follows:

	SAP report export to a file (*.txt; note: takes up to 20 min for some reports)
	start dnGrep and call the corresponding script (example see attachment)
		In the script we call the batch file
			- the batchfile searches the newest file in the directory and creates a backup file (*.sik in the same directory)
		in the script the single S&R are called up
		after the end the Orginaldatei over dnGrep was worked on and the Orginal state of the report's is in the same directory as *.sik file.
		
	Advantage: 
		Orginal version was saved before editing
		User can find the new file under the name of the SAP report (*.txt).
		User can process this file directly, e.g. import into Excel.
		
	Disadvantage:
		works only for the last current file
		cannot be used for mass processing of files (which would make sense)
	
[!CopyFile_TXT2SIK.cmd.txt](https://github.com/dnGrep/dnGrep/files/13162588/CopyFile_TXT2SIK.cmd.txt)
[SAP_ZZZP_pXXmP_SingelMatch_202310.gsc.txt](https://github.com/dnGrep/dnGrep/files/13162589/SAP_ZZZP_pXXmP_SingelMatch_202310.gsc.txt)

**Comment by doug24** [comment]
If I understand this correctly, you want to replace the batch file with dnGrep script commands and be able to process more than the most recent file.  In place of the batch file, the new part of the script might look like this:

```
set pathtomatch *.txt
set searchfor ""
search
copy -match (.*)\.txt -rename $1.txt.sik
```
Explanation:

1. Set the Path to match to text files - *.txt.
2. Set the Search For to empty string to list all *.txt files in the selected directory (the empty string currently has a bug and does not work as a script command).
3. Run the search to generate the file list.
4. For each file in the list, make a copy in the local directory renaming the copy using the search and replace regular expressions.

`copy` (not sure about the name or parameters yet) would be a new script command.  The first parameter is a regular expression to match on each file name, and the second parameter is a regular expression pattern to make the name of the copied file. In this example, SAP_ZZZP_pXXmP_1234.txt would be copied to SAP_ZZZP_pXXmP_1234.txt.sik.  If the match fails or if the copied file name already exists, skip making the copy.

Still thinking about this: There are not any other script commands with extended or variable argument lists.  But this command could have an `-overwrite true/false` parameter.  It could have an `-out` folder name to copy to another location, but there is already another command to do that.

**Comment by Alex-Kima** [comment]
Hi!
please help me
there are a lot of files like this
10221.1A.1.VPS
10221.1A.2.VPS
10221.1A.3.VPS
I need to open it, find the first word PANTONE in the text and add the numbers (Capture group) after it to the file name
But I can’t even copy and rename a file using this script. The result is a file with the same name.
```
set folder c:\vps
set pathtomatch *.vps
set searchfor "PANTONE[^\d]*(\d{2,8})."
search 
copy -match (.*)\.vps -rename $1\.bak -overwrite True -out c:\vps\bak
```
![1](https://github.com/dnGrep/dnGrep/assets/151007430/fe8cb2b4-eb80-46b2-b178-e8a7265d440f)
![2](https://github.com/dnGrep/dnGrep/assets/151007430/8227d0fc-35fa-4331-b198-acde8fd79c72)

what am I doing wrong? and is it even possible to add сapture group to the file name?

thank you!

**Reply by doug24** [reply]
The problem you are seeing is because the regular expression is case sensitive by default.  So `(.*)\.vps` does not match filenames with `.VPS` as the extension.  When the match expression does not match anything, the file doesn't get renamed. I probably should have set the case sensitive flag to false for this and will think about doing it for the next release. 

For now you can do it yourself my adding an [inline option](https://learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-options) to the beginning of the of the -match expression: `(?i)(.*)\.vps`  or just use `(.*)\.VPS` if the extensions are always uppercase.

One other note: you **do not** escape the dot in your replace expression - it should be just `$1.bak`.  With the backslash it will generate a file name of `10221.1A.1\.bak` and that will create a folder named `10221.1A.1` with a file named `.bak`

**Reply by Alex-Kima** [reply]
Thanks a lot! It worked.
But I still haven’t found how to add the found сapturegroup to a new file name. Is this possible with a script?
For example: 
old name 10221.1A.1.VPS  
new name 10221.1A.1-4975.VPS

**Reply by doug24** [reply]
That's not possible using the copy command.  In the copy command the capture groups for the -replace must come from the -match argument. 

It would be possible to do this, but not without some external code.  You could write the capture group text for each .VPS file out in a CSV report, then call a PowerShell script do the file renaming.  You would have to write the PowerShell script to read the CSV file, parse the line to extract the file name and Pantone number, create a new file name and rename the original file.

The CSV report will look like this (assuming all the Pantone numbers are the same in each file):
```
File Name,Line Number,String
C:\vps\10221.1A.1.VPS,,4975
C:\vps\10221.1A.3.VPS,,4842
C:\vps\10221.1A.2.VPS,,5235
```

Your script would look something like this (after setting and saving the Report Options):
```
set folder c:\vps
set pathtomatch *.vps
set searchfor "PANTONE[^\d]*(\d{2,8})."
search 
copy -match (.*)\.vps -rename $1\.bak -overwrite True -out c:\vps\bak
report csv c:\vps\captures.csv
run powershell c:\vps\rename.ps1
```

Here are the Report Options to get the report shown above (from the dnGrep main window - More - Report options...):

<img src="https://github.com/dnGrep/dnGrep/assets/17227248/46dbcba2-2f91-476f-b7fe-af644fa121d4" width="480"/>

**Reply by Alex-Kima** [reply]
Yes, that's exactly what I did)
I received a csv file and renamed the files using excel macro.
Thank you!
